### PR TITLE
fix content.find(object_provides) for and query

### PR DESCRIPTION
**Pull request by frisi**
this fixes https://github.com/plone/plone.api/issues/426

@mauritsvanrees this is my first PR on a package using towncrier. is the news file correctly formatted?

### CODE REVIEW

File: news/426.bugfix [hunk 1]
```diff
@@ -0,0 +1,2 @@
+- Fix querying `object_provides` for multiple interfaces using 'and' operator
```
Comment (mauritsvanrees) [inline news/426.bugfix]: Don't include a dash in front of the line in the news file.  `towncrier` will add this automatically when gathering all news snippets.
For the rest it is generally fine.   I would make it double quotes though (for literal/monospace) and add a dot at the end:

```
Fix querying ``object_provides`` for multiple interfaces using 'and' operator.
[fRiSi]
```

Comment (frisi) [reply news/426.bugfix]: thanks for your feedback - fixed

**Comment by frisi** [comment]
@mauritsvanrees i factored out the part that parses the query. though i personally did not have the feeling that content.find was overwhelmingly complex ;-)

code analysis still complains about the usage of `directlyProvides`. however the suggested replacement `zope.interface.provider` works on class level and we need to add the marker interface to an object...

**Comment by ale-rt** [comment]
Use `alsoProvides`! directlyProvides is evil http://blog.redturtle.it/2010/04/02 CC @keul

**Comment by frisi** [comment]
> Use `alsoProvides`! `directlyProvides` is evil http://blog.redturtle.it/2010/04/02 CC @keul

sure - i should have known that @ale-rt. thanks!

**Comment by frisi** [comment]
@jenkins-plone-org please run jobs

**Comment by frisi** [comment]
the errors in jenkins look unrelated. restarting the tests
@jenkins-plone-org please run jobs

**Comment by frisi** [comment]
@mauritsvanrees i'm running out of ideas here:

integration tests on jenkins seem flaky. previously 3 failed, later just one. i'm running them once again.
@jenkins-plone-org please run jobs

flake8 tests on travis complain about modules/lines i did not even touch:

```flake8 runtests: commands[2] | flake8 src setup.py
src/plone/api/portal.py:45:42: W504 line break after binary operator
src/plone/api/portal.py:50:38: W504 line break after binary operator
src/plone/api/tests/test_doctests.py:35:34: W504 line break after binary operator
src/plone/api/tests/test_doctests.py:36:22: W504 line break after binary operator
src/plone/api/tests/test_doctests.py:37:26: W504 line break after binary operator
src/plone/api/tests/test_env.py:28:21: W605 invalid escape sequence '\d'
src/plone/api/tests/test_env.py:28:25: W605 invalid escape sequence '\.'
src/plone/api/tests/test_env.py:28:27: W605 invalid escape sequence '\d'
src/plone/api/tests/test_env.py:28:39: W605 invalid escape sequence '\d'
src/plone/api/tests/test_env.py:28:44: W605 invalid escape sequence '\d'
src/plone/api/tests/test_env.py:28:50: W605 invalid escape sequence '\d'
src/plone/api/tests/test_env.py:28:56: W605 invalid escape sequence '\.'
src/plone/api/tests/test_env.py:28:61: W605 invalid escape sequence '\d'
src/plone/api/tests/test_content.py:508:50: W504 line break after binary operator
src/plone/api/tests/test_content.py:524:48: W504 line break after binary operator
src/plone/api/tests/test_content.py:538:54: W504 line break after binary operator
src/plone/api/tests/test_content.py:551:42: W504 line break after binary operator
src/plone/api/tests/test_content.py:596:50: W504 line break after binary operator
src/plone/api/tests/test_content.py:625:54: W504 line break after binary operator
src/plone/api/tests/test_content.py:644:56: W504 line break after binary operator
src/plone/api/tests/test_content.py:691:40: W504 line break after binary operator
src/plone/api/tests/test_content.py:702:44: W504 line break after binary operator
src/plone/api/tests/test_content.py:722:50: W504 line break after binary operator
src/plone/api/tests/test_content.py:732:42: W504 line break after binary operator
ERROR: InvocationError for command '/home/travis/build/plone/plone.api/.tox/flake8/bin/flake8 src setup.py' (exited with code 1)
```

**Comment by mauritsvanrees** [comment]
All Jenkins tests pass now.  Lots of changes in packages came in today, so you might have been unlucky getting a run in between two related merges.  Also, I have the feeling that Jenkins has more failures when  lots of jobs are run concurrently, especially more failures with robot tests that lose the connection with the browser or the server.

Last Travis run on master was apparently a few months ago and it was green then.  Since then, probably a new version of `flake8` or of several plugins was released, and this gives rise to new errors.  Joy! ;-)

Let me see if I can fix a few on your branch.

**Comment by mauritsvanrees** [comment]
Okay, I have fixed `W605 invalid escape sequence` by changing a regular expression to use a raw string.

But the others are tricky. They are all `W504 line break after binary operator`. When I do a fix, the W504 goes away, but is replaced by its reverse: `W503 line break before binary operator`. So we should tell Travis/tox to ignore one or both of them.

@gforcada Do you have a recommendation here?  See [`tox.ini`](https://github.com/plone/plone.api/blob/frisi/fix-object-provides-426/tox.ini#L57).

Meanwhile:

@jenkins-plone-org please run jobs

**Comment by mauritsvanrees** [comment]
When I look at the explanation of [W503](https://lintlyci.github.io/Flake8Rules/rules/W503.html) and [W504](https://lintlyci.github.io/Flake8Rules/rules/W504.html), the first should be ignored, and the second should be followed. For example, best practice would be:

```
income = (gross_wages
          + taxable_interest)
```

But most of all, as [pep8 says](https://www.python.org/dev/peps/pep-0008/#should-a-line-break-before-or-after-a-binary-operator): it should be consistent.

**Comment by mauritsvanrees** [comment]
I have pushed a fix. Locally `flake8` is happy now.

@jenkins-plone-org please run jobs

**Comment by mauritsvanrees** [comment]
And now the Jenkins QA report fails on W503... See the [comment above](https://github.com/plone/plone.api/pull/427#commitcomment-32291545).

@gforcada Maybe Jenkins QA should ignore W503 as well?

**Comment by mauritsvanrees** [comment]
Hah, Travis at least passes now!

**Comment by jensens** [comment]
I propose to adopt black code formatter in future to avoid such time consuming and pointless problems/work.

**Comment by frisi** [comment]
thanks @mauritsvanrees.

having a less picky code-analysis would be a great help here.as @jensens points out these changes are pointless and very time consuming.

https://black.readthedocs.io/en/stable/ looks good but i guess replacing the tools used for code analysis throughout plone will require a plip ;-)

hope @gforcada agrees add some more Warning codes to the ignores.

**Comment by gforcada** [comment]
tl;dr; yes, let's merge as it is and find the correct flake8 configuration, be it in tox.ini, setup.cfg, or wherever that might be :+1: 

At first, I was not fond on the idea of black, I guess I'm getting old :older_man: :rofl: , replacing code-analysis by a formatter increases the consistency but misses all the analysis :sweat_smile: 

Fast forward though, we adopted **black+isort+fine tuned flake8 and plugins** and the results are great! :smile: :tada: :confetti_ball: 

But as it was suggested, a PLIP should be written, and more than a PLIP and someone running black over 300 packages, we need a way to ensure we do not regress. The least benefit we can get from blackifying our code is that we keep adding at every second pull request a `format code according to black` commits.

**Comment by frisi** [comment]
thanks for your feedback @mauritsvanrees @gforcada and @jensens

**Comment by frisi** [comment]
and @ale-rt of course :-P
